sensor: mcp9808: adding ability to change the resolution of the sensor through kconfig

sdaglish opened the pull request:
The current MCP9808 driver does not set the resolution for the sensor, instead using the default resolution of 0.0625C.

In some cases, it is better to use a lower resolution (for example to reduce the time required to acquire a new sample).

This PR adds the ability to change the resolution of the temperature measurements at compile time.

As the resolution is set during the MCP9808 init function, the i2c write functions have been moved from mcp9808_trigger.c to mcp9808.c.

Since the resolution register is only 8 bits, an 8bit write function has been created, and the original 16bit write function renamed to show the difference.

MaureenHelm commented on drivers/sensor/mcp9808/mcp9808.h:
```diff
@@ -1,4 +1,4 @@
-/*
+*
```
this doesn't look right...

sdaglish replied:
This was fixed in the latest push.

MaureenHelm commented on drivers/sensor/mcp9808/mcp9808.h:
```diff
@@ -16,11 +16,11 @@
+#define MCP9808_REG_UPPER_LIMIT         0x02
+#define MCP9808_REG_LOWER_LIMIT         0x03
+#define MCP9808_REG_CRITICAL            0x04
+#define MCP9808_REG_TEMP_AMB            0x05
```
The whitespace was fine, why do you want to change it?

MaureenHelm commented on samples/sensor/mcp9808/boards/nucleo_l053r8.overlay:
```diff
@@ -0,0 +1,14 @@
+		int-gpios = <&gpioa 8 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
+		label = "MCP9808";
+	};
+};
```
Please add this board to the `sample.yaml`

sdaglish replied:
The board has now been added to `sample.yaml`

MaureenHelm commented on drivers/sensor/mcp9808/Kconfig:
```diff
@@ -30,6 +30,24 @@ config MCP9808_TRIGGER_OWN_THREAD
 
 endchoice
 
+choice
+	prompt "MCP9808 temperature sensor relolution"
+	default MCP9808_MEAS_RES_0_0625C
+
+config MCP9808_MEAS_RES_0_5C
+	bool "0.5C temperature resolution"
+
+config MCP9808_MEAS_RES_0_25C
+	bool "0.25C temperature resolution"
+
+config MCP9808_MEAS_RES_0_125C
+	bool "0.125C temperature resolution"
+
+config MCP9808_MEAS_RES_0_0625C
+	bool "0.0625C temperature resolution"
+
+endchoice
```
We have been migrating toward defining these as dts properties instead of Kconfigs, as this allows different selections per instance.

sdaglish replied:
Is there an example of this being used in one of the sample projects? I'm struggling to find an example of this implementation to work from.

MaureenHelm replied:
See #30317

MaureenHelm reviewed (changes requested):
Thanks for the PR!

Please move the resolution selection from a Kconfig choice to a device tree property.

I don't think the whitespace commit is necessary.

sdaglish committed:
sensor: mcp9808: Added dts resolution to nucleo_l031k6 board overlay

MCP9808 sensor driver now uses DTS to change the resolution of the
sensor.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

sdaglish committed:
sensor: mcp9808: Add new overlay board for nucleo_l053r8 board

Introduce a dts board overlay on the nucleo_l053r8 board as this
board was used for testing addition of resolution setting on the
mcp9808 sensor driver.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

sdaglish committed:
sensor: mcp9808: Adding temperature resolution to MCP9808 sensor driver

During the driver init, the function will set the sensor resolution
based on the driver's dts variable "resolution"

The driver's device tree has been updated to include the value
"resolution".

The default is set to the highest resolution of 0.0625C.

Moved mcp9808_reg_write from mcp9808_trigger.c to mcp9808.c

This allows resue of the same function in both the trigger and
resolution functions.

Function name changed to xxx_16bit to distinquish it from the 8
bit write function that will added.

Signed-off-by: Steven Daglish <s.c.daglish@gmail.com>

sdaglish reviewed (commented):
Thanks for the example commit. I have now moved the resolution selection into the dts instead of the kconfig.

MaureenHelm commented on dts/bindings/sensor/microchip,mcp9808.yaml:
```diff
@@ -18,3 +18,14 @@ properties:
+    resolution:
+      type: int
+      required: false
+      default: 3
```
Is there a reason you picked this default value? Is it the reset value in the sensor register? Please document where the default came from in the description.

sdaglish replied:
This is the default, power-on, value. This information has now been documented.

MaureenHelm reviewed (changes requested):
> I have now moved the resolution selection into the dts instead of the kconfig.

Thank you. This part looks good.

MaureenHelm reviewed (approved):
Looks good, thank you.

Please review and try to fix the buildkite failures.

avisconti commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -82,6 +115,12 @@ int mcp9808_init(const struct device *dev)
 
+	rc = mcp9808_set_temperature_resolution(dev, cfg->resolution);
+	if (!rc) {
+		LOG_DBG("Could not set the resolution of mcp9808 module");
```
Should not be LOG_ERR()?

sdaglish replied:
Agreed and amended.

avisconti commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
+}
+
+static int mcp9808_set_temperature_resolution(const struct device *dev,
+					      uint8_t resolution)
```
I'm wondering if this is the correct indentation. I think 'uint8_t' should be aligned under 'const'. But I also see that this would break the other indentations in the driver. I don't know what to suggest, if leaving as it is in accordance with the rest of the driver, or to fix it.

sdaglish replied:
Happy with either convention. Small personal preference to align `uint8_t` with `const`. That is how it's showing on my device, but I need to change the indentation correctly. I'll update this and re-push.

avisconti commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
+int mcp9808_reg_write_8bit(const struct device *dev, uint8_t reg,
+			   uint8_t val)
+{
+	const struct mcp9808_data *data = dev->data;
```
Is data struct really 'const'?

sdaglish replied:
I believe that in this instance it should be `const`. `i2c_write` requires `const struct device *dev` as first parameter.

avisconti commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
+int mcp9808_reg_write_16bit(const struct device *dev, uint8_t reg,
+			    uint16_t val)
+{
+	const struct mcp9808_data *data = dev->data;
```
Is data struct really 'const'?

sdaglish replied:
I believe that in this instance it should be `const`. `i2c_write` requires `const struct device *dev` as first parameter.

gmarull replied:
driver `data` is never `const`, devices are.

sdaglish replied:
Updated

avisconti reviewed (changes requested):
LGTM, just minor things

gmarull commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
+	const struct mcp9808_config *cfg = dev->config;
+	uint8_t buf[3] = {
+		reg,
+		val >> 8,       /* big-endian register storage */
```
use facilities in `byteorder.h`

sdaglish replied:
`const` removed from data.

I'm struggling to figure out why using facilities from `byteorder.h` will make this small section any better. It's already clear what the section is doing.

I've read through 'byteroder.h' and even after changing the order, I would still need to manually place the different bytes within the array.

gmarull replied:
you can use `sys_put_be16` to insert the value

sdaglish replied:
Sorry, but I'm still not seeing this. I can use the following to include `sys_put_be16`, but I still don't see it as being any better:

`uint8_t buff[3] = {reg};
sys_put_be16(val, &buf[1]);`

It compiles and runs, but I'm not seeing the benefit.

sdaglish replied:
Or

`uint8_t buff[3];
buff[0] = reg;
sys_put_be16(val, &buf[1]);`

gmarull replied:
I think it is better to use system facilities when available, they'll usually be tested and in some cases optimized.

sdaglish replied:
Updated to use `sys_put_be16`

gmarull commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -35,6 +35,39 @@ int mcp9808_reg_read(const struct device *dev, uint8_t reg, uint16_t *val)
+{
+	const struct mcp9808_data *data = dev->data;
+	const struct mcp9808_config *cfg = dev->config;
+	uint8_t buf[3] = {
```
shouldn't it be `2`?

sdaglish replied:
That is correct and amended.

gmarull commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -78,10 +111,16 @@ int mcp9808_init(const struct device *dev)
 	}
 
+	rc = mcp9808_set_temperature_resolution(dev, cfg->resolution);
+	if (!rc) {
```
This eval will be true for rc == 0, which is not error. Have you tested these changes?

sdaglish replied:
This was changed at request: https://github.com/zephyrproject-rtos/zephyr/pull/31758#discussion_r568488344

Please confirm if this is to be set as an error or not. If not, I'll put it back to log debug.

gmarull replied:
the problem is not `LOG_ERR`, this branch will be executed if rc == 0, i.e. operation succeeds.

sdaglish replied:
After testing with logging enabled, I fully see what you mean. This will be updated.

gmarull commented on drivers/sensor/mcp9808/mcp9808.c:
```diff
@@ -97,9 +136,10 @@ static const struct mcp9808_config mcp9808_cfg = {
-		    &mcp9808_data, &mcp9808_cfg, POST_KERNEL,
-		    CONFIG_SENSOR_INIT_PRIORITY, &mcp9808_api_funcs);
+		      &mcp9808_data, &mcp9808_cfg, POST_KERNEL,
+		      CONFIG_SENSOR_INIT_PRIORITY, &mcp9808_api_funcs);
```
is this change needed to align?

sdaglish replied:
This is the output from the default uncrustify. I presume this change is correct.

gmarull commented on samples/sensor/mcp9808/sample.yaml:
```diff
@@ -6,3 +6,4 @@ tests:
     tags: sensors
     depends_on: i2c
     filter: dt_compat_enabled("microchip,mcp9808")
+    platform_allow: nucleo_l053r8
```
not sure if this is strictly needed

sdaglish replied:
Not necessary, but I think good to add.

Addition was requested here: https://github.com/zephyrproject-rtos/zephyr/pull/31758#discussion_r566399165

gmarull commented on samples/sensor/mcp9808/boards/nucleo_l053r8.overlay:
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) 2019, Linaro Ltd.
```
Copyright should likely be adjusted (doesn't seem to be copied?)

sdaglish replied:
Fixed.

gmarull reviewed (approved):
Added a couple of minor comments, but the rest LGTM. Thanks!

avisconti reviewed (approved):
LGTM now
